### PR TITLE
Icons v2

### LINKED ISSUE #873: Roll out new icon set
## Context

The brand design team is working on a new set of custom icons for SumUp. These will replace the existing icons in this package. The icons will come in filled + outlined versions and 4 different sizes (16x16, 24x24, 48x48, and 64x64 pixels). Here are the [WIP guidelines](https://sumupteam.atlassian.net/wiki/spaces/DES/pages/1758527666/Icons).

## Tasks

- Promote the use of `@sumup/icons`. The more it's used in our apps already, the easier the rollout will be.
- Map the new icons to the existing icon names.

### PR DESCRIPTION
**Pull request by robinmetral**
Closes #873 

## Purpose

Introduces new brand icons.

## Approach and changes

- Keep v1 icons in the `v1` directory, they will continue to be statically deployed
- Update the icons script to generate components from `v2` with new size props (`16`, `24`) and a new default (`24`)
- Upgrade icons in Circuit (this will be a major—`@sumup/circuit-ui@4.0.0`
- Add a single codemod to handle the icon changes (see discussions 👇)

## Definition of done

* [x] Development completed
* [x] Reviewers assigned
* [x] Unit and integration tests
* [x] Meets minimum browser support
* [x] Meets accessibility requirements

### CODE REVIEW

File: packages/circuit-ui/components/Notification/Notification.tsx [hunk 1]
```diff
@@ -59,9 +55,9 @@ const colorMap = {
 } as const;
 
 const iconMap = {
-  success: CircleCheckmarkFilled,
-  error: CircleCrossFilled,
-  warning: CircleWarningFilled,
+  success: Confirm,
+  error: Alert,
+  warning: Notify,
```
Comment (connor-baer) [inline packages/circuit-ui/components/Notification/Notification.tsx:60]: We should align these variant names in the new notification components.

**Review by connor-baer — COMMENTED** [review]
The new names are so much more semantic, I love it!

File: packages/circuit-ui/components/Notification/Notification.tsx (hunk 1 above)
Comment (robinmetral) [reply packages/circuit-ui/components/Notification/Notification.tsx:60]: Excellent point! Wonder if v4 should also rename them in the legacy notification components?

Comment (connor-baer) [reply packages/circuit-ui/components/Notification/Notification.tsx:60]: Maybe, up to you.

Comment (robinmetral) [reply packages/circuit-ui/components/Notification/Notification.tsx:60]: In the end I'd prefer keeping this for the new notification components

File: .changeset/cyan-rivers-sleep.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@sumup/circuit-ui': patch
```
Comment (robinmetral) [inline .changeset/cyan-rivers-sleep.md:2]: Note: remember to add the missing changesets for the major (I will write a migration guide in a separate PR)

File: packages/circuit-ui/cli/migrate/__tests__/migrate.spec.js [hunk 3]
```diff
@@ -54,6 +54,9 @@ defineTest('theme-border-radius');
 defineTest('button-default-size');
 defineTest('label-prop-names');
 
+// v4
+defineTest('icons-v2');
```
Comment (robinmetral) [inline packages/circuit-ui/cli/migrate/__tests__/migrate.spec.js:58]: I kept the transforms in a single codemod, thought it was not helpful to split and make developers run several of them, especially since the transforms are so closely related

File: packages/circuit-ui/cli/migrate/icons-v2.ts [hunk 4]
```diff
@@ -0,0 +1,309 @@
+/**
```
Comment (robinmetral) [inline packages/circuit-ui/cli/migrate/icons-v2.ts:1]: Not really happy about this codemod, open to suggestions. I find it very verbose and hard to follow. But I think it also does the job 🤷 I'll test it locally on one of our apps before merging

File: packages/circuit-ui/components/Popover/__snapshots__/Popover.spec.tsx.snap [hunk 5]
```diff
@@ -223,19 +223,16 @@ exports[`Popover styles should render popover on auto 1`] = `
         <svg
           class="circuit-2"
           fill="none"
-          height="16"
-          viewBox="0 0 16 16"
-          width="16"
+          height="24"
+          size="24"
```
Comment (robinmetral) [inline packages/circuit-ui/components/Popover/__snapshots__/Popover.spec.tsx.snap:227]: Example of the `size` making it into the DOM. Will need to investigate a bit further, because this doesn't happen consistently, and I think that we're destructuring `size` as expected in the icon components (will add another comment on the relevant spot below 👇)

File: packages/circuit-ui/components/SideNavigation/components/DesktopNavigation/__snapshots__/DesktopNavigation.spec.tsx.snap [hunk 6]
```diff
@@ -654,15 +650,14 @@ exports[`DesktopNavigation styles should render without secondary links 1`] = `
                 fill="none"
                 height="24"
                 role="presentation"
+                size="24"
```
Comment (robinmetral) [inline packages/circuit-ui/components/SideNavigation/components/DesktopNavigation/__snapshots__/DesktopNavigation.spec.tsx.snap:653]: (here's another one)

File: packages/circuit-ui/components/SideNavigation/components/MobileNavigation/__snapshots__/MobileNavigation.spec.tsx.snap [hunk 7]
```diff
@@ -933,15 +926,14 @@ exports[`MobileNavigation styles should render without secondary links 1`] = `
                       fill="none"
                       height="24"
                       role="presentation"
+                      size="24"
```
Comment (robinmetral) [inline packages/circuit-ui/components/SideNavigation/components/MobileNavigation/__snapshots__/MobileNavigation.spec.tsx.snap:929]: (and another)

File: packages/circuit-ui/components/Sidebar/components/Aggregator/__snapshots__/Aggregator.spec.tsx.snap [hunk 8]
```diff
@@ -245,16 +245,14 @@ exports[`Aggregator styles should render with disabled state styles and match th
       <svg
         fill="none"
         height="24"
+        size="24"
```
Comment (robinmetral) [inline packages/circuit-ui/components/Sidebar/components/Aggregator/__snapshots__/Aggregator.spec.tsx.snap:248]: 🙄

File: packages/circuit-ui/components/TopNavigation/components/UtilityLinks/__snapshots__/UtilityLinks.spec.tsx.snap [hunk 9]
```diff
@@ -122,55 +122,14 @@ exports[`UtilityLinks styles should match the snapshot 1`] = `
           fill="none"
           height="24"
           role="presentation"
+          size="24"
```
Comment (robinmetral) [inline packages/circuit-ui/components/TopNavigation/components/UtilityLinks/__snapshots__/UtilityLinks.spec.tsx.snap:125]: 🙄

File: packages/icons/scripts/web.ts [hunk 10]
```diff
@@ -85,7 +85,7 @@ function buildComponentFile(component: Component): string {
       ${sizeMap.join('\n')}
     }
 
-    export const ${component.name} = ({ size = 'small', ...props }) => {
+    export const ${component.name} = ({ size = '24', ...props }) => {
```
Comment (robinmetral) [inline packages/icons/scripts/web.ts:88]: This is where we're destructuring the `size`. Since we're only spreading `props`, it should not be in the DOM 🤔

**Review by robinmetral — COMMENTED** [review]
I'll mark this ready for review so that you, dear reviewer, can take an early look, but there's a couple of things left to do here.

1. Figure out why this `size` attribute ends up in the DOM sometimes
2. Write the changesets
3. Test the codemods locally on a real application

After this is done and all your PR review comments are addressed, I'll merge this into `next` (recreated the branch and already entered prerelease mode) and we can start migrating a few apps before releasing a stable `circuit-ui@4.0.0` and `icons@2.0.0`.

File: packages/circuit-ui/components/IconButton/IconButton.tsx [hunk 11]
```diff
@@ -40,7 +40,7 @@ const sizeStyles = (size: IconButtonProps['size'] = 'giga') => (
 ): SerializedStyles => {
   const sizeMap = {
     kilo: theme.spacings.byte,
-    giga: theme.spacings.mega,
+    giga: theme.spacings.kilo,
```
Comment (connor-baer) [inline packages/circuit-ui/components/IconButton/IconButton.tsx:43]: Could we automatically set the icon size based on the button size on line 58?

File: packages/icons/scripts/web.ts (hunk 10 above)
Comment (connor-baer) [reply packages/icons/scripts/web.ts:88]: This code is only used when multiple sizes are available. If there's only one size, the code on L69-L71 is used. I'd suggest destructuring the size there, too.

Comment (robinmetral) [reply packages/icons/scripts/web.ts:88]: 👀 nice catch, that's why

File: packages/circuit-ui/cli/migrate/icons-v2.ts [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+    .find(j.Identifier)
+    .filter((nodePath) => {
+      const hasLegacyIconName = nodePath.node.name === oldIconName;
+      if (hasLegacyIconName && productName) {
+        console.warn(
+          [
+            `The "${oldIconName}" icon has been renamed to "${newIconName}",`,
+            `and should only be used in the context of the ${productName}`,
+            `product/feature from now on.`,
+            `If you have doubts about your use of the icon, contact #design-system.`,
```
[the remark this answers is not in the document]
Comment (robinmetral) [reply packages/circuit-ui/cli/migrate/icons-v2.ts:203]: Good point, will update in each warning

File: packages/circuit-ui/cli/migrate/icons-v2.ts [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+  JSCodeshift,
+  Collection,
+  JSXAttribute,
+  StringLiteral,
+} from 'jscodeshift';
+
+import { findImportsByPath, findStyledComponentNames } from './utils';
+
+const RENAMED_ICONS = [
+  ['AddItem', 'AddItems', 'Item Catalog'],
```
[the remark this answers is not in the document]
Comment (robinmetral) [reply packages/circuit-ui/cli/migrate/icons-v2.ts:27]: Yes, good idea 👍 It's actually a tiny bit more complicated than that 😅 in this case it's `[oldName: string, newName: string, productName?: string]` (we're warning if product-specific icons are used because eventually they should be "reserved" to the said products)

File: packages/circuit-ui/components/IconButton/IconButton.tsx (hunk 11 above)
Comment (robinmetral) [reply packages/circuit-ui/components/IconButton/IconButton.tsx:43]: Do you mean pass the relevant `size` down to the children (a single icon) so that it doesn't need to be added manually?

Can we do this while avoiding this? 👇 

```tsx
<IconButton>
  {({ size }) => <Icon size={size} />}
</IconButton>
```

File: packages/circuit-ui/cli/migrate/icons-v2.ts [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+
+  if (!componentImport) {
+    return;
+  }
+
+  root.find(j.Identifier).filter((nodePath) => {
+    const hasLegacyIconName = nodePath.node.name === oldIconName;
+    if (hasLegacyIconName) {
+      const defaultMessage =
+        'Copy it locally to finish the migration, and request a new icon for your use case in #design-system.';
```
[the remark this answers is not in the document]
Comment (robinmetral) [reply packages/circuit-ui/cli/migrate/icons-v2.ts]: Done in 699fd0c

File: packages/circuit-ui/cli/migrate/icons-v2.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (robinmetral) [reply packages/circuit-ui/cli/migrate/icons-v2.ts]: Done in 699fd0c

File: packages/circuit-ui/cli/migrate/icons-v2.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (robinmetral) [reply packages/circuit-ui/cli/migrate/icons-v2.ts:28]: Done in 699fd0c

File: packages/icons/scripts/web.ts (hunk 10 above)
Comment (robinmetral) [reply packages/icons/scripts/web.ts:92]: Is this what you. had in mind? 3ec788e

Comment (connor-baer) [reply packages/icons/scripts/web.ts:92]: Yes, that's exactly it.

File: packages/circuit-ui/components/IconButton/IconButton.tsx (hunk 11 above)
Comment (connor-baer) [reply packages/circuit-ui/components/IconButton/IconButton.tsx:43]: Yes. The IconButton already clones its child to add the `role: 'presentation'` prop, you can add the size there as well.

Comment (robinmetral) [reply packages/circuit-ui/components/IconButton/IconButton.tsx:43]: Nice, I missed that! Done in 35969d1.

The only thing I'm wondering is, what if the icon doesn't exist in the relevant size? (e.g. an IconButton with a "AddProduct" which only exists in 24px) Would TS throw an error? No right? So we'd just have an icon with the wrong size, and probably could spot it in development

Comment (connor-baer) [reply packages/circuit-ui/components/IconButton/IconButton.tsx:43]: That's true, TypeScript wouldn't catch that 😕

**Review by connor-baer — APPROVED** [review]
💙

**Review by amelako — APPROVED** [review]
❣️🚀

File: packages/circuit-ui/components/IconButton/IconButton.tsx (hunk 11 above)
Comment (robinmetral) [reply packages/circuit-ui/components/IconButton/IconButton.tsx:44]: ☝️ actually it would be filtered out just like for the other issue we had with the `size` prop in the DOM, I guess

**Review by connor-baer — APPROVED** [review]
I love the addition of `chalk` for the codemods, very nice! 

Your attention to detail is exemplary 🥇
